buildSrc/src/main/java/org/elasticsearch/gradle/release/BreakingChangesGenerator.java

```diff
@@ -0,0 +1,197 @@
+        private final String majorMinor;
+        private final String majorDotMinor;
+        private final String nextMajor;
+
+        private VersionStrings(Version version) {
+            this.majorMinor = String.valueOf(version.getMajor()) + version.getMinor();
+            this.majorDotMinor = version.getMajor() + "." + version.getMinor();
+            this.nextMajor = (version.getMajor() + 1) + ".0";
+        }
+    }
+
+    private void generateBreakingChanges(List<ChangelogEntry> entries, VersionStrings versionStrings) {
+        final Map<String, List<ChangelogEntry.Breaking>> breakingChangesByArea = entries.stream()
+            .map(ChangelogEntry::getBreaking)
+            .filter(Objects::nonNull)
+            .collect(Collectors.groupingBy(ChangelogEntry.Breaking::getArea, TreeMap::new, Collectors.toList()));
+
+        if (breakingChangesByArea.isEmpty()) {
+            return;
+        }
+
+        out.println();
+        out.println();
+
+        List.of(
+            "[discrete]",
+            "[[breaking-changes-" + versionStrings.majorDotMinor + "]]",
+            "=== Breaking changes",
+            "",
+            "The following changes in {es} " + versionStrings.majorDotMinor + " might affect your applications",
+            "and prevent them from operating normally.",
+            "Before upgrading to " + versionStrings.majorDotMinor + " review these changes and take the described steps",
+            "to mitigate the impact.",
+            "",
+            "NOTE: Breaking changes introduced in minor versions are",
+            "normally limited to security and bug fixes.",
+            "Significant changes in behavior are deprecated in a minor release and",
+            "the old behavior is supported until the next major release.",
+            "To find out if you are using any deprecated functionality,",
+            "enable <<deprecation-logging, deprecation logging>>."
+        ).forEach(out::println);
+
+        breakingChangesByArea.forEach((area, breakingChanges) -> {
+            out.println();
+
+            final boolean hasNotableChanges = breakingChanges.stream().anyMatch(ChangelogEntry.Breaking::isNotable);
+            if (hasNotableChanges) {
+                out.println("// tag::notable-breaking-changes[]");
+            }
+
+            out.println("[discrete]");
+            out.println(
+                "[[breaking_" + versionStrings.majorMinor + "_" + area.toLowerCase(Locale.ROOT).replaceAll("[^a-z0-9]+", "_") + "]]"
+            );
+            out.println("==== " + area);
+
+            breakingChanges.forEach(breaking -> {
+                out.println();
+                out.println("[[" + breaking.getAnchor() + "]]");
+                out.println("." + breaking.getTitle());
+                out.println("[%collapsible]");
+                out.println("====");
+                out.println("*Details* +");
+                out.println(breaking.getBody().trim());
+                out.println("====");
```

Comment:
For some changes, there is also an Impact section that describes what action you need to take.